### ISSUE TITLE
How to check user auth to write a data, eg. a post document

### ISSUE THREAD

**Issue by EryouHao** [fillipvt/with-faunadb-abac-auth#1]
For example, I have a posts collection, a user login and to update a post, how to check if the post is create by this user? must use a query anyway? Or can be checked from other places?

thanks.

**Comment by fev4** [comment]
Okay so, let me see if I understand it.

- A user logs in: a mutation - **loginuser**
- The user gets the current posts owned by them: a query that gets all posts - **allPosts**
- The user updates one post from the current list: a mutation - **updatePost**
- Or the user creates one more post and adds it to the list:  a mutation - **createPost**

Okay, so, given the aforementioned, you should have a token that identifies this specific user when the user logged in. This token is very important because it defines to what functions, indexes and document the user has access to and what type of [privileges](https://docs.fauna.com/fauna/current/security/abac.html#privileges) they have over those resources.

If you have a type definition for `Post` similar to this:
````
type Post {
  title: String!
  content: String!
  owner: User!
}
````

Then during post creation you just have to attach any identifiable information to the **createPost** mutation (any unique value within the `User` type) in order to guarantee, that only logged in users are able to create a document. This is the first type of block you can have, by using an `owner` field.

Now if you want to guarantee that only users who own a `Post` can update it, you do that at two stages.

At the `read` action, similar [to this](https://github.com/fillipvt/with-faunadb-abac-auth#employee-role). Pay special attention to these couple of lines:

````
Equals(
  Select(["data", "owner"], Var("file")),
  Var("userRef")
)
````

This is how you set the second limit for accessing the `Post`s so that only a user who owns said `Post` is able to read it, and then from the front-end be able to update it.

Then, you can set a limit at the `write` step, [here the docs](https://docs.fauna.com/fauna/current/security/abac.html#predicates) mention that the `write` operations have access to both the `old data` and the `new data` and since both **have to have** an `owner` field you can do an equality check there too.

With the previous three limits, at the, create, read and write steps I think you are pretty safe in in your normal CRUD operations.

So in short, you guarantee this by using predicate functions inside your role definitions in order to define the [Attribute-based access control (ABAC)](https://docs.fauna.com/fauna/current/security/abac.html) environment your app has. And by defining some sort of non-null ownership (like the `owner` field) for the files you want protected.

Let me know if this is clear or not!
